python/ql/src/experimental/semmle/python/frameworks/Django.qll

```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+        private DataFlow::Node expressions(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importNode("django.db.models.expressions")
+          or
+          t.startInAttr("expressions") and
+          result = models()
+          or
+          exists(DataFlow::TypeTracker t2 | result = expressions(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `django.db.models.expressions` object. */
+        DataFlow::Node expressions() { result = expressions(DataFlow::TypeTracker::end()) }
+
+        /** Gets a reference to the `django.db.models.expressions.RawSQL` class. */
+        private DataFlow::Node classRawSQL(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importNode("django.db.models.expressions.RawSQL")
+          or
+          t.startInAttr("RawSQL") and
+          result = expressions()
+          or
+          exists(DataFlow::TypeTracker t2 | result = classRawSQL(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `django.db.models.expressions.RawSQL` class. */
+        DataFlow::Node classRawSQL() { result = classRawSQL(DataFlow::TypeTracker::end()) }
+      }
+    }
+  }
+
+  /** A call to the `django.db.connection.cursor.execute` function. */
+  private class DbConnectionExecute extends SqlExecution::Range, DataFlow::CfgNode {
+    override CallNode node;
+
+    DbConnectionExecute() { node.getFunction() = django::db::execute().asCfgNode() }
+
+    override DataFlow::Node getSql() {
+      result.asCfgNode() in [node.getArg(0), node.getArgByName("sql")]
+    }
+  }
+
+  /** A call to the `annotate` function on a model using a `RawSQL` argument. */
+  private class ObjectsAnnotate extends SqlExecution::Range, DataFlow::CfgNode {
+    override CallNode node;
+    CallNode raw;
+
+    ObjectsAnnotate() {
+      node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
+      raw in [node.getArg(0), node.getArgByName(_)] and
+      raw.getFunction() = django::db::models::classRawSQL().asCfgNode()
+    }
+
+    override DataFlow::Node getSql() { result.asCfgNode() = raw.getArg(0) }
+  }
+
+  /** A call to the `raw` function on a model. */
+  private class ObjectsRaw extends SqlExecution::Range, DataFlow::CfgNode {
+    override CallNode node;
+
+    ObjectsRaw() { node.getFunction() = django::db::models::objects_attr("raw").asCfgNode() }
+
+    override DataFlow::Node getSql() { result.asCfgNode() = node.getArg(0) }
+  }
+
+  /** A call to the `raw` function on a model. */
+  private class ObjectsExtra extends SqlExecution::Range, DataFlow::CfgNode {
+    override CallNode node;
+
+    ObjectsExtra() { node.getFunction() = django::db::models::objects_attr("extra").asCfgNode() }
+
+    override DataFlow::Node getSql() {
+      result.asCfgNode() =
+        [node.getArg([0 .. 5]),
+            node.getArgByName(["select", "where", "params", "tables", "order_by", "select_params"])]
+    }
+  }
```

Comment:
If you do expand functionality of the parts of the API we talked about not trying to fully support, I think you need to go the full distance and add tests as well :stuck_out_tongue:

Besides this, it is actually safe to pass in user-controlled data to `params` and `select_params`

Reply:
I will take those out, although there might be the issue of having quoted placeholders. But since I have done nothing with that anywhere yet, it seems `params` and `select_params` should indeed be ignored for now.

Reply:
:+1: to me, `params` and `select_params` are considered safe, but your SQL template can be unsafe it if includes `"where username = '%s'"`, since you shouldn't add the quoting yourself (and we should indeed handle this in the same for `cursor().execute(SQL, params...)`)